flink-ai-flow/wiki_doc/Tutorial.md

```diff
@@ -0,0 +1,383 @@
+
+For `validate` job, we only config its job type to be `python`.
+
+For `predict` job, we set its job type to be `flink`, which means this job is a flink job. In addition, to use flink, we need to set some flink-specific properties including `run_mode` (cluster or local) and `flink_run_args`. For the `flink_run_args`, we follow the list format of `yaml` to add the args such as the `-pyexec`.
+
+## Writing the Workflow Code
+
+Now, let's start coding to implement workflow described in Target section
+
+### Importing Modules
+A workflow Flink AI Flow is just a Python script. Let’s create `tutorial_workflow.py` and start by importing the libraries we will need.
+
+
+```python
+import os
+
+# Entry of accessing Flink AI Flow API
+import ai_flow as af
+
+# Enum class for marking different stages of our workflow
+from ai_flow.model_center.entity.model_version_stage import ModelVersionEventType
+
+# Util functions provided by ai_flow
+from ai_flow.util.path_util import get_file_dir
+
+# User-defined processors; we will introduce them in tutorial_processors.py
+from tutorial_processors import DatasetReader, ModelTrainer, ValidateDatasetReader, ModelValidator, Source, Sink, \
+    Predictor
+
+```
+
+###  Defining a Training Job
+In our design, the workflow in Flink AI Flow is a DAG(Directed Acyclic Graph) or to be more specifically, it is a [AIGraph](). Each node in the graph is an [AINode](), which contains a processor. Users should write their custom logic in the processor. 
+
+In the AIGraph, nodes are connected by 2 types of edges. The first one is named as `DataEdge` which means the destination node depends on the output of the source node. The other is `ControlEdge` which means the destination node depends on the control conditions from source node. We will dive deeper into this kind of edges later.
+
+For now, let's concentrate on the set of AINodes connected by only data edges. Such a set of nodes and data edges between them constitutes a sub-graph. This sub-graph, together with the predefined job config in workflow config yaml, is mapped to a Job by Flink AI Flow framework.
+
+So, to summarize, each Flink AI Flow job is made up of a job config, some AINodes and the data edges between those nodes. 
+
+These concepts seem to be a little difficult, but do not worry. Let's look at some code snippets, and you will find it pretty easy to define a job with provided API of Flink AI Flow.
+
+In the following codes, we define a training job of our workflow:
+```python
+    artifact_prefix = af.current_project_config().get_project_name() + "."
+    # Training of model
+    with af.job_config('train'):
+        # Register metadata of training data(dataset), `uri` refers to the file path
+        train_dataset = af.register_dataset(name=artifact_prefix + 'train_dataset',
+        																		uri=DATASET_URI.format('train'))
```

Comment:
where is DATASET_URI defined?

Reply:
Add DATASET_URI in ### Importing Modules section.